core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -428,6 +444,20 @@ public CompletableFuture<Optional<Throwable>> acknowledge(String memberId, List<
         Throwable throwable = null;
         lock.writeLock().lock();
         List<InFlightState> updatedStates = new ArrayList<>();
+        boolean canMoveStartOffset = false;
+        Iterator<AcknowledgementBatch> iterator = acknowledgementBatch.iterator();
+        AcknowledgementBatch firstBatch = iterator.next();
+        // The Share Partition Start Offset can be moved after acknowledgement is complete when the following 3 conditions are true -
+        // 1. When the cachedState is not empty
+        // 2. When the acknowledge request includes the acknowledgement data for startOffset
+        // 3. When the acknowledgement type for the base offset is either ACCEPT or REJECT
+        if (!cachedState.isEmpty()) {
+            if (firstBatch.baseOffset == startOffset &&
+                    (firstBatch.acknowledgeType == AcknowledgeType.ACCEPT ||
+                            firstBatch.acknowledgeType == AcknowledgeType.REJECT)) {
+                canMoveStartOffset = true;
+            }
+        }
```

Comment:
I think the computation should be the last thing in the method when all records have been successfully acked. That's the time when we shall be calling persister and then this functionality can trigger. If we can also move the functionality to a different method then it will ease the refactoring as well later on while integrating with persister.